File: MaWPictures/src/main/java/us/mikeandwan/photos/ui/photos/PhotoViewPager.java
package us.mikeandwan.photos.ui.photos;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.MotionEvent;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;
import com.github.chrisbanes.photoview.PhotoView;


// http://stackoverflow.com/questions/11306037/how-to-implement-zoom-pan-and-drag-on-viewpager-in-android
// https://raw.githubusercontent.com/chrisbanes/PhotoView/master/sample/src/main/java/uk/co/senab/photoview/sample/HackyViewPager.java
public class PhotoViewPager extends ViewPager {
    private final PublishSubject<Integer> _photoSelectedSubject = PublishSubject.create();
    private boolean _enabled = true;


    public PhotoViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);

        addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                // do nothing
            }

            @Override
            public void onPageSelected(int position) {
                _photoSelectedSubject.onNext(position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                // do nothing
            }
        });
    }


    @Override
    public boolean onInterceptTouchEvent(MotionEvent arg0) {
        try {
            return _enabled && super.onInterceptTouchEvent(arg0);
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
            return false;
        }
    }


    public Observable<Integer> onPhotoSelected() {
        return _photoSelectedSubject.hide();
    }


    public boolean isEnabled() {
        return _enabled;
    }


    public void setEnabled(boolean enabled) {
        _enabled = enabled;
    }


    public void rotateImage(int direction) {
        PhotoView pv = (PhotoView) findViewWithTag(getCurrentItem());

        pv.setRotation(pv.getRotation() + (direction * 90));
    }
}

File: MaWPhotos/src/main/java/us/mikeandwan/photos/ui/photos/PhotoViewPager.kt
package us.mikeandwan.photos.ui.photos

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.viewpager.widget.ViewPager
import com.github.chrisbanes.photoview.PhotoView
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

// http://stackoverflow.com/questions/11306037/how-to-implement-zoom-pan-and-drag-on-viewpager-in-android
// https://raw.githubusercontent.com/chrisbanes/PhotoView/master/sample/src/main/java/uk/co/senab/photoview/sample/HackyViewPager.java
class PhotoViewPager(context: Context?, attrs: AttributeSet?) : ViewPager(
    context!!, attrs
) {
    private val _photoSelectedSubject = PublishSubject.create<Int>()
    private var _enabled = true
    override fun onInterceptTouchEvent(arg0: MotionEvent): Boolean {
        return try {
            _enabled && super.onInterceptTouchEvent(arg0)
        } catch (e: IllegalArgumentException) {
            e.printStackTrace()
            false
        }
    }

    fun onPhotoSelected(): Observable<Int> {
        return _photoSelectedSubject.hide()
    }

    override fun isEnabled(): Boolean {
        return _enabled
    }

    override fun setEnabled(enabled: Boolean) {
        _enabled = enabled
    }

    fun rotateImage(direction: Int) {
        val pv: PhotoView = findViewWithTag(currentItem)
        pv.rotation = pv.rotation + direction * 90
    }

    init {
        addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) {
                // do nothing
            }

            override fun onPageSelected(position: Int) {
                _photoSelectedSubject.onNext(position)
            }

            override fun onPageScrollStateChanged(state: Int) {
                // do nothing
            }
        })
    }
}